[FLINK-2055] Implement Streaming HBaseSink

delding opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

delding commented:
Hi, this is my second PR to Flink, any comments or suggestions will be very appreciated :-)

delding commented:
BTW, failing tests are results of this PR

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/pom.xml:
```diff
@@ -0,0 +1,68 @@
+    <parent>
+        <groupId>org.apache.flink</groupId>
+        <artifactId>flink-streaming-connectors</artifactId>
+        <version>1.1-SNAPSHOT</version>
```
We've recently just bumped version to 1.2-SNAPSHOT.

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,100 @@
+import java.io.IOException;
+import java.util.List;
+
+public class HBaseSink<IN> extends RichSinkFunction<IN> {
```
Since this is public facing, we should have Javadocs for this class.

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,100 @@
+	private HBaseMapper<IN> mapper;
+	private boolean writeToWAL = true;
+
+	public HBaseSink(String tableName, HBaseMapper<IN> mapper) {
```
Will need Javadocs for the constructor too :)

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,100 @@
+		this.mapper = mapper;
+	}
+
+	public HBaseSink writeToWAL(boolean writeToWAL) {
```
Again, Javadoc for public facing interfaces.

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseMapper.java:
```diff
@@ -0,0 +1,97 @@
+		private byte[] qualifier;
+		private byte[] value;
+		private long timestamp;
+		private Long increment;
```
I would make these `final`.

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,66 @@
+import java.io.IOException;
+import java.util.List;
+
+public class HBaseClient implements Closeable {
```
Can you also add a simple Javadoc for this class?

tzulitai commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,100 @@
+package org.apache.flink.streaming.connectors.hbase;
+
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
```
Can you try removing the Guava dependencies? In Flink we try to avoid using Guava because of dependency issues.

tzulitai commented:
@delding Thank you for working on this contribution. Since I'm not that familiar with the HBase client API, I've only skimmed through the code for now. Will try to review in detail later.

Btw, since there are no tests for this PR yet, can you describe how you tested it? An example of using the `HBaseSink` will also be helpful (we'd also need to add documentation for the sink, like the other connectors).

delding commented:
Hi @tzulitai Thank you for all the comments :-) I will update the PR and add an example in next a few days

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseMapper.java:
```diff
@@ -0,0 +1,97 @@
+		private Long increment;
+
+		public HBaseColumn(byte[] family, byte[] qualifier, byte[] value) {
+			this(family, qualifier, value, -1);
```
Should this be -1? Ideally if there is no ts HBase will assign the server based timestamp.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseMapper.java:
```diff
@@ -0,0 +1,97 @@
+		}
+
+		public Long getIncrement() {
+			return increment;
```
May be we can add an API explicity for increment?  Instead of having a boolean to identify increment?

ramkrish86 replied:
The reason to say this was even Append is also a Mutation.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,100 @@
+			if (column.isStandard()) {
+				mutation = new Put(rowKey);
+				if (column.getTs() == -1) {
+					((Put) mutation).addColumn(column.getFamily(), column.getQualifier(), column.getValue());
```
So the -1 is handled here. So ignore that comment for ts set to -1.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,100 @@
+		List<Mutation> mutations = Lists.newArrayList();
+		for (HBaseMapper.HBaseColumn column : mapper.columns(value)) {
+			Mutation mutation;
+			if (column.isStandard()) {
```
Better to created Enums for PUT, DELETE, INCREMENT and APPEND and handle all of them here? so that Table#batch() will work for all. But Deletes are much more complicated because there are ROW deletes, COL deletes , family deletes, deletes with specific versions.

delding replied:
Thanks for these helpful comments :-)

delding commented:
Hi, I have updated this PR, adding javadoc and an example as @tzulitai advised and created Enums for PUT, DELETE, INCREMENT and APPEND as @ramkrish86 advised.

delding commented:
BTW, the example runs well on my machine when I set hbase-client version to 2.0.0-SNAPSHOT, but travis-ci build failed due to "Failed to collect dependencies at org.apache.habse:hbase-client:jar:2.0.0-SNAPSHOT" error. So I changed hbase dependency version back to 1.2.2 which unfortunately has a bug ( https://issues.apache.org/jira/browse/HBASE-14963 ) that makes the example fail to run. The bug is fixed for version 1.3.0+ and 2.0.0

ramkrish86 commented:
Thanks for the update @delding . I can check this PR more closely tomorrow my time. Had a glance at it.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+					append.add(action.getFamily(), action.getQualifier(), action.getValue());
+					break;
+
+				// If there are multiple DELETE_ROW actions, only the first one is served
```
May be take the last one rather than the first one.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addFamily(action.getFamily(), action.getTs());
```
If action.getTs() == -1 use the delete.addFamily(byte[] fam) version. which deletes all columns of the specified family.

delding replied:
This case is already covered when calling MutationActionList#addDeleteFamily without timestamp, then ts field is assigned with Long.MAX_VALUE which will cause all columns with the specified family to be deleted.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addColumns(action.getFamily(), action.getQualifier(), action.getTs());
```
If action.getTs() == -1, call delete.addColumns(byte[] fam, byte[] qual) so that all versions of a specified column can be deleted.

delding replied:
Same as above, Long.MAX_VALUE will be assigned to ts field if calling MutationActionList#addDeleteColumns without timestamp.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addColumn(action.getFamily(), action.getQualifier(), action.getTs());
```
if action.getTs()== -1 then you can call delete.addColumn(byte[] fam, byte[] qual) so that latest version of the specified column can be removed. But for all the above you may have to add so many enums. May be you can check with the Flink community as to how to proceed here. Should we add all the delete related support here or can be done in this JIRA only.

delding replied:
I forget this case, will add code to handle it as I did for DeleteFamily and DeleteColumns, which don't require extra enum type.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+	 * @return this
+	 */
+	public MutationActionList addPut(byte[] family, byte[] qualifier, byte[] value, long timestamp) {
+		actions.add(new MutationAction(family, qualifier, value, timestamp, 0, MutationAction.Type.PUT));
```
Is it not better to add your mutation - I mean create a new Mutation object and add it to the MutationList. Now this will add to actions and then you have to call newMutatiionList again. Am not sure if this looks easy for the end user.

delding replied:
The design is to let users define actions they want to perform on a single row given a input record and return a MutationActionList for it. HBaseSink will then call newMutatiionList only once given an input and send associated Mutations to an HBase table. What do you think?

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+	 * @return a list of HBase {@link Mutation}s
+	 */
+	public List<Mutation> newMutationList(byte[] rowKey) {
+		return newMutationList(rowKey, true);
```
What will happen if I try to call this directly?  I think this API is not needed.  Always make writeToWAL as default and I think you can have a default state boolean for that. If needed let the user override it.

delding replied:
There is already a writeToWAL field which default to be true in HBaseSink class. I agree this method is not needed as it is expected to be called by HBaseSink. Will remove it.

ramkrish86 replied:
> And would it be even better to make newMutationList(byte[] rowKey, boolean writeToWAL) a non-public method?

Exactly. That was my point. My below comments also was to make things cleaner. So MutationActionList is exposed to the user. It should only have getMutations() or createMutations() and addXXX(). So when the user creates a MuationActionList using addXXX APIs in it. Then he should do getMutations() or createMutations()(add suitable javadoc) saying this API gives a list of mutations that can be sent to HBase. In this method you need to do the logic of what newMutationList() does. Do not expose getActions() also. I was suggesting to do this in MutationAction but fine in MutationActionList as only that is going to be exposed now. Better rename it to MutationActions. Mark MutationAction inner class Private. And none of those Getters in the inner class should be exposed to User. If needed you can create it as a separate class also.
And one important thing, the rowKey() in HBaseMapper decides the row for which these mutations are getting created. If this is set to null then there is no MutationActionList possible at all. So you should ensure that the rowKey if null then the mapper should fail. If not just don''t add any rowKey explicitly - Anyway actions(IN value) API  accepts the rowkey that should be enough. Inside the impl - if the value is found null throw exception and add a javadoc about it. It should be enough and will be simple too.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+		return this;
+	}
+
+	static class MutationAction {
```
Will it not be better to expose MutationAction - and allow the user to create a MutationAction and then just add them to the MutationList. MutationAction can have all the different flavours of addXXX() APIs?  The return type of each will be a MutationAction. That will get added to a MutationActionList and so the getActions() in mutation Action will just return the actions added to it. What do you think? Related to the HBaseSink implementing RichSinkFunction I think people from Flink community can answer you the best.

delding replied:
addXXX() APIs can cover all the flavors of MutationActions (including delete entire family, delete all columns and delete the most recent column you mentioned above) and help users create them. I think it would be easier to use than let users create these actions themselves.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseClient.java:
```diff
@@ -0,0 +1,69 @@
+
+	public HBaseClient(org.apache.hadoop.conf.Configuration hbConfig, String tableName) throws IOException {
+		connection = ConnectionFactory.createConnection(hbConfig);
+		table = connection.getTable(TableName.valueOf(tableName));
```
Have you tested when there is no table available with that name? May be you need to have connect() API like in CassandraSink so that if that fails then there is no point in continuing with others and gives you a clear API specifically for connecting.

delding replied:
You are right, I will do it.

ramkrish86 replied:
Not able to see the connect method. May be it is not added?

delding commented:
Hi @ramkrish86 , thanks for your reviews and detailed comments :-)
I will update this PR soon.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java:
```diff
@@ -0,0 +1,104 @@
+	private transient HBaseClient client;
+	private String tableName;
+	private HBaseMapper<IN> mapper;
+	private boolean writeToWAL = true;
```
My bad. I missed this. It is there as you said.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java:
```diff
@@ -0,0 +1,371 @@
+					break;
+
+				case DELETE_COLUMN:
+					if (!rowIsDeleted) {
```
I don't think you need to handle this. Let the user create the Deletes, it is upto the hbase to handle all these combinations. Any specific reason you wanted to handle it here?

delding replied:
I would be more than happy not to handle combination of Deletes :-). Will update.

rmetzger commented:
@delding: Thanks a lot for opening a pull request for a streaming hbase connector to Flink.
Would you be okay with contributing the connector to Apache Bahir, instead of Apache Flink?
The repository is located here: https://github.com/apache/bahir-flink

The Flink community recently decided to move some of our connectors to Apache Bahir [1][2].
If you want, I can also help migrating the pull request over to Bahir.
At Bahir, I'll help with the review and make sure it gets merged.
Also, once the first Flink connectors are in Bahir, I'm proposing (to the Bahir community) to create a first Bahir-flink release to get the code out. My goal is to have the code out well ahead of the next Flink release.

[1] http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Move-Redis-and-Flume-connectors-to-Apache-Bahir-and-redirect-contributions-there-td13102.html
[2] http://mail-archives.apache.org/mod_mbox/flink-dev/201608.mbox/%3CCAGr9p8CAN8KQTM6%2B3%2B%3DNv8M3ggYEE9gSqdKaKLQiWsWsKzZ21Q%40mail.gmail.com%3E

delding commented:
Hi @rmetzger , Thanks for your comment. I am totally ok with contributing the connector to Apache Bahir :-)

rmetzger commented:
Cool. Let me know if you have any questions or if you need any help.
I'll keep an eye on this PR and the Bahir-flink repository.

delding commented:
Hi, I have updated the PR based on @ramkrish86 's comments.
1. rename MutationActionList to MutationActions which has a public method createMutations that  
   return HBase Mutations
2. make MutationAction a private inner class
3. DeleteColumn action can now delete the latest version of the specified column
4. check if rowkey is null when create Mutations
5. hand over the combinations of Delete logic to HBase server
And I checked if table doesn't exist in HBase in which case a table not found exception will be thrown

ramkrish86 commented:
How should I view your latest commit ?  Am not able to see the changes that you have mentioned above. Thanks @delding

delding commented:
Hi @ramkrish86 , I squashed the commits before pushing.. Sorry about making it unable to see those recent changes. If you could still review this PR when you have time?

ramkrish86 commented:
@delding 
Am not able to see the recent changes at all. Under 'commit' tab it shows 2 commits. If I open the merge commit I am not able to see the changes that you are saying here. Can you point me to the right one?

delding commented:
Hi @ramkrish86 , I squashed my recent commit into the one you reviewed before, then merged the upstream and pushed --force the commits which is why you only see two commits. The first commit actually includes those changes based on your comments.

ramkrish86 commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseMapper.java:
```diff
@@ -0,0 +1,46 @@
+	 * @param value
+	 * @return row key
+	 */
+	byte[] rowKey(IN value);
```
Fine with all. But how do you related this rowKey and the one to be passed to createMutations?  May be will it be better to have actions(byte[] rowkey, IN value)? and add a getRowKey() API that returns byte[]? This can be used for createMutations?

ramkrish86 replied:
Rest looks good to me. I think naming of functions am not sure if it can be fine tuned. Am not very good at naming. And also FLINK team may have some naming conventions. I think having a connect() method and the above comment are the main things from my side. If it is fine, then it is upto Flink team to review this. Thanks @delding .

delding replied:
The assumption is that for each input value there will be only one corresponding row to be modified. Users define how to extract a row key and actions to be performed on this row key from a single input value. In other words, the row key returned from rowKey(IN value) will be passed to createMutations(byte[] rowKey, boolean writeToWAL) method called on the instance returned from actions(IN value).

delding replied:
I did forget to add a connect method. Will add one shortly.

ramkrish86 replied:
> In other words, the row key returned from rowKey(IN value) will be passed to createMutations(byte[] rowKey, boolean writeToWAL) method called on the instance returned from actions(IN value). How do you think?

Am not entirely convinced on having a rowKey API and use that in createMutations. Since HBase is going to allow row key as byte[] I think there is no harm in exposing an argument with byte[] which represents the rowkey. Hence i was of that opinion.

delding replied:
Hi @ramkrish86 , if we have 
MutationActions mutationActions  = HBaseMapper#actions(rowkey, value);
Then, the rowkey information would be included in mutationActions object being created. So when we call mutationActions. createMutations() there would be no need to pass rowkey as an argument, which I think is totally fine. But what I'm confused is how do we derive rowkey from the input value? Should we pass input value as an argument to getRowKey() API you mentioned? If so, getRowKey() API would be the same as rowkey(IN value) except a name change.

ramkrish86 replied:
> If so, getRowKey() API would be the same as rowkey(IN value) except a name change.

May be yes.  that is better? Ok. Let's see what others has to say. Can you ping @rmetzger for getting this in and for further reviews?

delding replied:
Hi @ramkrish86 , I was trying to and connect() method as you suggested and to explicitly test if table exists as following:
public void connect(String tableName) throws IOException {
		connection = ConnectionFactory.createConnection(hbConfig);
		TableName name = TableName.valueOf(tableName);
		try (Admin admin = connection.getAdmin()) {
			if (!admin.tableExists(name)) {
				throw new RuntimeException("Table " + tableName + " doesn't exist!");
			}
		}
		table = connection.getTable(name);
	}
But once the method admin.tableExists(name) is added, running my example will throw exception: org.apache.hadoop.hbase.regionserver.NoSuchColumnFamilyException: Column family table does not exist in region hbase:meta,,1.1588230740 in table 'hbase:meta'
I've never run into this exception in my project that uses hbase client. Do you know the reason for this exception?

delding replied:
I am planing update this PR based on your comment and then ping @rmetzger for further review.

fhueske commented on flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActions.java:
```diff
@@ -0,0 +1,362 @@
+ *  Each {@link MutationAction} can create an HBase {@link Mutation} operation type
+ *  including {@link Put}, {@link Increment}, {@link Append} and {@link Delete}.
+ */
+public class MutationActions {
```
Maybe I am missing something, but wouldn't it be easier to have all `addX` methods implemented in the HBaseMapper and let them directly return a `Mutation`?

fhueske commented on flink-streaming-connectors/flink-connector-hbase/src/test/java/org/apache/flink/streaming/connectors/hbase/example/HBaseSinkExample.java:
```diff
@@ -0,0 +1,118 @@
+		}
+
+		@Override
+		public MutationActions actions(Tuple3<String, Integer, Integer> value) {
```
I think this example is a bit too complicated. Can you change it to two or three different actions?

delding commented:
Hi @fhueske , I have updated this PR to address your comments. In this change, only one MutationActions and two ArralyList are needed for entire stream, MutationActions is now resettable. But I replaced MutationActions actions(IN value) with void actions(IN value, MutationActions actions) instead of void actions(IN value, List<Mutation> actions) as you suggested. Because in this case, user can still utilize MutationActions's API to handle Mutations creation logic which makes user easier to code. I also added a connect() method as @ramkrish86 suggested, but didn't explicitly check existence of table in the code, because once Admin#tableExists is added a NoSuchColumnFamilyException will be thrown during running the example. I have no idea why this happens... For now, if table doesn't exist, a table not existence Exception will be thrown until HBaseClient#send() is called.

fhueske commented:
Hi @delding, I'm sorry that I did not mention this earlier, but I just noticed that the `HBaseSink` does not implement any logic for checkpointing and fault-tolerance. 

The checkpointing logic of a sink and the guarantees of the external data store define what kind of emission guarantees a sink function can offer (exactly once, at-least once). In order to make this PR mergable, we need to figure out what kind of guarantees it provides under which assumptions / in which scenarios. As an example see the [documentation of the Cassandra sink](https://ci.apache.org/projects/flink/flink-docs-master/dev/connectors/cassandra.html). This is very important because users will rely on the stated guarantees.

I am not very familiar with HBase and what kind of write semantics and transnational guarantees it provides. Can you describe which kind of guarantees an `HBaseSink` can achieve and how it would be integrated with Flink's Checkpointing mechanism? 

Thanks, Fabian

fhueske commented:
I also noticed that the PR pulls in an HBase SNAPSHOT dependency. This is not possible. We cannot compile against a development version. #2330 updates the version of the batch TableInputFormat to HBase 1.1.2. I think we should use the same version here.

delding commented:
Hi @fhueske , in HBase writes to a single row have ACID guarantee. The exactly once semantic can be implemented the way CassandraSink did, storing input records in a state backend and flushing to HBase upon receiving a checkpoint barrier. One thing that might be a concern is the order of execution of these writes are not defined when making such a batch call. In other words, a write of an earlier record could be observed later, but this could be also true even sending each write immediately. So what do you think if implement fault tolerance the same way as Cassandra Sink?

delding commented:
I can change back to HBase 1.1.2. The reason of using 2.0.0-SNAPSHOT is because this bug (https://issues.apache.org/jira/browse/HBASE-14963 ). It's only fixed for version 1.3.0+ and 2.0.0 and my example has problem of running because of it.

ramkrish86 commented:
` #2330 updates the version of the batch TableInputFormat to HBase 1.1.2. I think we should use the same version here.`
Valid point. But is it possible to upgrade to even newer version like 1.1.6 or the branch1.2 series?

ramkrish86 commented:
I agree with @delding here. 

>    * Method that does a batch call on Deletes, Gets, Puts, Increments and Appends.
   * The ordering of execution of the actions is not defined. Meaning if you do a Put and a
   * Get in the same {@link #batch} call, you will not necessarily be
   * guaranteed that the Get returns what the Put had put.

is the javadoc for batch API.  Anyway here we don't do get operation but still the order of execution among mutations is not guarenteed.
Regarding 

>  storing input records in a state backend and flushing to HBase upon receiving a checkpoint barrier.

Where is this being done in Flink. Just for understanding.

delding commented:
Hi @ramkrish86 , CassandraTupleWriteAheadSink extends GenericWriteAheadSink that deals with the checkpoint mechanism

fhueske commented:
Thanks for looking into the Cassandra connector. Maybe @zentol (who implemented the Cassandra connector) can comment how an HBase sink can be made fault-tolerant and what code could be reused for that.

Idempotent (put, delete) and non-idempotent modifications (increment, append) have different implications on fault-tolerance semantics. I think it would be a good idea to separate both types of modifications to make this difference more clear.

I would be OK updating the version update to HBase 1.1.6 or 1.2.x under the condition that these versions are backwards compatible to 1.1.2. @nielsbasjes, would that be OK for you as well? We have an ongoing discussion to merge batch and streaming connectors. So this code would end up in the same module as the batch `TableInputFormat` and we should keep the HBase versions synced.

zentol commented:
Don't you loose any guarantees regarding order of mutations the moment you use asynchronous updates anyway?

The WriteAheadSink should only be used if you want to deal with non-deterministic programs or want to send data in atomic mini-batches and must rely on the order of elements. Otherwise there are much simpler solutions.

If you do idempotent updates the only thing you have to do is write the data into HBase, and make sure that every update sent for a given checkpoint is acknowledged before it is regarded as complete. If you don't acknowledge them you lose at-least-once guarantees. This scheme does not provide exactly-once *delivery* guarantees, however at any given point in time the table would be in a state as if the updates were only sent once. This is the same guarantee that we provide for Cassandra.

For non-idempotent updates the thing gets a lot more difficult.

If you can fire an entire checkpoint as a single atomic batch you just won the lottery, as you can use the above scheme and a small auxiliary table to track completed checkpoints per sink subtask.

if you can't do that you will have to use system-specific features/guarantees to engineer a solution that provides exactly-once guarantees. Versioning, rollbacks, unique ID's; something that either allows you to revert the table to a clean state or track precisely which updates were applied and sent the remaining updates.

delding commented:
Thanks for very detailed and helpful comments, let me work on it :-)

nielsbasjes commented:
Yes, as long as everything is compatible with HBase 1.1.2 its fine for me.

ramkrish86 commented:
> This scheme does not provide exactly-once delivery guarantees, however at any given point in time the table would be in a state as if the updates were only sent once. This is the same guarantee that we provide for Cassandra.

@zentol - so for Cassandra also it is atleast-once guarentee is what we provide? So even if the mutations happens once again since the operations are idempotent we don't have any issues with such duplicates? Am I getting it right here?

zentol commented:
@ramkrish86 that is correct.

nielsbasjes commented:
@fhueske Should the TableInputFormat be updated to use the HBase 1.1.2 api also? It would make things a bit cleaner.

fhueske commented:
@nielsbasjes We would break compatibility with older HBase versions. Before we do that, we should start a discussion on the user and dev mailing lists.

nielsbasjes commented:
@fhueske: What is "older" ?
I would like a clear statement about the (minimal) supported versions of HBase. 
I would see 1.1.x as old enough, or do you see 0.98 still required?

fhueske commented:
@nielsbasjes I want to avoid to exclude a version that is still in use by existing Flink users. I do not have in insight in which HBase versions currently in use. If (basically) everybody is on 1.1.x I am definitely in favor of bumping the version and API for Flink 1.2.0. Let's move the discussion to the dev and user mailing lists.

ramkrish86 commented:
@delding 
Do you have uses cases with Append/Increment? 
I think with the batch() API we are not sure of the order of execution of the batch() in the hbase server but still it would help us to achieve the atleast-once guarantee as @zentol  mentioned here.

delding commented:
Hi @ramkrish86 , I'm thinking replace batch() with mutateRow() because it provides atomic ordered mutations for a single row, but it only supports Put and Delete which should be fine since only Put and Delete are idempotent, this way we can implement Put and Delete without using WriteAheadSink (in case of deterministic processing). What do you think? 

Regarding Append and Delete, as HBase doesn't support distributed transaction across multiple rows, we might need WriteAheadSink and figure out a way to roll back table to the last checkpoint state. I'm thinking about this right now.

So it might make sense to have two HBaseSinks, one for Put/Delet, the other for Append/Delete and non-deterministic processing.

ramkrish86 commented:
With in a single row you need guarantee of order of execution? I agree Append/Increment or non-idempotent in certain failure cases but there is a Nonce generator for that. I should say I have not used that anyway to know the exact pros and cons of that. RowMutations atleast it supports only Put/Delete so in that angle you can be sure that for now we don't support Append/Increments.

>  we might need WriteAheadSink and figure out a way to roll back table to the last checkpoint state

This will be tricky. Your hbase table rollback would mean that you may have to issue Deletes here so that the previous mutations are hidden.

delding commented:
I agree it might be an overkill. But in case of having an sink that only supports Put/Delete, it would be better to have ordered execution than not, after all HBase has this API so there could be some use case that needs order guarantee.

ramkrish86 commented:
Ok. Then it should be clearly documented now. that the sink supports only Puts/Deletes. So in future can the sink be updated with new APIs? I don't know the procedure here in case  new APIs have to be added in the sink. Since the sink is not part of the FLINK distro it can be enhanced or modified. Just wanted to know. May be as a first step make things simple
->Support only Puts/ Delets
-> You want order to be guaranteed then go with RowMutations.
-> clearly document what the sink does now.
-> Pls verify if it is ok to update the sink to support other APIs in future.
Anyway this will give you atleast-once guarentee.

fhueske commented:
Thanks for these discussions @delding and @ramkrish86. I think this is getting a bit hard to follow. What do you think about writing a design doc that describes the supported HBase features and consistency guarantees of the sink? The design doc does not need to be a full fetched [FLIP](https://cwiki.apache.org/confluence/display/FLINK/Flink+Improvement+Proposals), just a document to share and discuss. I think this would make it easier for others to follow and join the discussion. I also find that explaining designs helps to improve them.

nragon commented:
Hi :)

I needed to use hbase as sink so i decided to take a look at this pull a use it.
Current changes that might be interesting: Like hbase connector for dataset, it's possible to define the configuration in other to connect, for instance, to a remote hbase master. In [HBaseSink](https://github.com/delding/flink/blob/16ad4b2ac13567ceba7bacf15e4698fb4ce17c53/flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/HBaseSink.java) the `HBaseConfiguration.create()` method should receive this configurations set by constructor (or other method)
Hope this makes sense.

Thanks

nragon commented:
Any updates on this sink?

fhueske commented:
Hi @nragon, the PR seems to be stale.
Are you interested in contributing a streaming HBase sink?

nragon commented:
Hi, I'm testing one my self with a third party library, http://opentsdb.github.io/asynchbase.
I'm following a simliar approach as cassandra sink. Testing it as we speak.
Thanks

ramkrish86 commented:
@delding 
Are you still working on this?
@nragon 
Let me know how can be of help here? I can work on this PR too since I have some context on the existing PR though it is stale.

delding commented:
Hi @ramkrish86 , sorry I don't have bandwidth working on it..

nragon commented:
Currently trying to fix some throttling issues, reported here https://github.com/OpenTSDB/asynchbase/issues/162

fpompermaier commented:
anyone working on this? HBase streaming sink would be a very nice addition...

nragon commented:
I've made a custom solution which works for my use cases. Notice that the code attached is not working because it's only a skeleton.
This prototype uses asynchbase and tries to manage throttling issues as mentioned above. The way I do this is by limiting requests per client by 1000 (also configurable, if you want, depending on hbase capacity and response), and skipping records after reaching that threshold. Every record skipped is updated according with system timestamp, always keeping the most recent skipped record for later updates.
Now, in my use case I always use a keyby -> reduce before sink, which keeps the aggregation state, meaning that every record invoked by hbase sink will have the last aggregated value from your previous operators. When all requests are done `pending == 0` I compare the last skipped record with the last requested record, if the skipped timestamp is less than the requested timestamp means that hbase has the last aggregation.
There is plenty of room for improvments, i just did'nt have the time.

[HBaseSink.txt](https://github.com/apache/flink/files/1014991/HBaseSink.txt)

ramkrish86 commented:
I can take this up and come up with a design doc. Reading thro the comments here and the final decision points I think only puts/deletes can be considered idempotent. But increments/decrements cannot be considered to be idempotent. We may need to two types of sink then one sink which supports the puts/deletes and the other one where we need to support non-idempotent ops.
Coming to @nielsbasjes issue of not able to flush the buffered mutator - should you always call bufferedmutator#flush() on every checkpoint call?

nragon commented:
From my point of view, my sample works fine (my use case)

ramkrish86 commented:
@nragon 
I agree. But your use case does it have increments/appends?

nragon commented:
No, only puts. Aggregation is coming from a reduce which by itself aggregates, keeping recent values on hbase, more like snapshots I think. During our analysis sending data to hbase was only reliable when working with aggregates otherwise a huge amount of backpressure will come from region servers. Full dumps are more likely to work on elasticsearch or hdfs than hbase. So, because Flink does an amazing job at keeping states why bother to overload region servers with so many requests? Again, our point of view, of course.

ramkrish86 commented:
So I think at a first level let us have put/delete mutations alone for Streaming ? Since am not aware of how flink users are currently interacting with HBase not sure on what HBase ops should be supported. From your case it seems puts/deletes are the common cases.

nragon commented:
Indeed, I'm also not aware of how users use it. I would take a similiar approach than c*, already in made for streaming, seems to use a existing async api from datastax. But again, I couldn't find accurate feedback regarding backpressure. If this is an issue, then working with Flink states is a must, imo.

nielsbasjes commented:
FYI:
In close relation to this issue I submitted an enhancement at the HBase side to support these kinds of usecases much better:
https://issues.apache.org/jira/browse/HBASE-19486
https://issues.apache.org/jira/browse/HBASE-19673

Simply put: You can now specify a max time the records may be kept in the BufferedMutator. If they exceed this value then a flush will be done automatically.

aljoscha commented:
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.
